File: tcfunct.cpp
#include <iostream>
#include <memory>
#include <deque>
#include <algorithm>
#include <chrono>

template<typename T>
class Stack {
    struct Node {
        T value;
        std::shared_ptr<Node> next;

        Node(T val, std::shared_ptr<Node> nxt) : value(val), next(nxt) {}
    };
    std::shared_ptr<Node> topNode;

public:
    Stack() : topNode(nullptr) {}

    Stack push(T value) const {
        return Stack(std::make_shared<Node>(value, topNode));
    }

    Stack pop() const {
        if (!topNode) return *this;
        return Stack(topNode->next);
    }

    size_t getMemorySize() const {
        size_t size = 0;
        auto node = topNode;
        while (node) {
            size += sizeof(*node);
            node = node->next;
        }
        return size;
    }

private:
    Stack(std::shared_ptr<Node> node) : topNode(node) {}
};

template <typename Func>
void measureTime(const std::string& operation, Func&& func) {
    auto start = std::chrono::high_resolution_clock::now();
    func();
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double, std::milli> duration = end - start;
    std::cout << operation << " took " << duration.count() << " ms" << std::endl;
}

void testStackOperations(size_t numElements) {
    Stack<int> stack;

    std::cout << "\nTesting " << numElements << " elements:\n";

      measureTime("Push operation", [&]() {
        for (size_t i = 0; i < numElements; ++i) {
            stack = stack.push(i);
        }
    });
    std::cout << "Memory usage after push: " << stack.getMemorySize() << " bytes" << std::endl;

        measureTime("Pop operation", [&]() {
        for (size_t i = 0; i < numElements; ++i) {
            stack = stack.pop();
        }
    });

}

int main() {
    testStackOperations(10000);
    testStackOperations(20000);
    testStackOperations(30000);

    return 0;
}

File: tcfunct.hs
import Data.Time.Clock (getCurrentTime, diffUTCTime)
import Control.DeepSeq (NFData, deepseq)
import GHC.Generics (Generic)

data Stack a = EmptyStack | StackNode a (Stack a) deriving (Show, Generic)
instance NFData a => NFData (Stack a)

push :: a -> Stack a -> Stack a
push x s = StackNode x s


pop :: Stack a -> Stack a
pop EmptyStack = EmptyStack
pop (StackNode _ s) = s


estimateMemorySize :: Stack a -> Int -> Int
estimateMemorySize EmptyStack _ = 0
estimateMemorySize (StackNode _ s) elementSize = elementSize + estimateMemorySize s elementSize


measureTime :: String -> IO a -> IO a
measureTime description action = do
    start <- getCurrentTime
    result <- action
    end <- getCurrentTime
    putStrLn $ description ++ " took " ++ show (diffUTCTime end start)
    return result

forceEvaluation :: NFData a => a -> IO ()
forceEvaluation x = x `deepseq` return ()


testStackOperations :: Int -> IO ()
testStackOperations numElements = do
    let elementSize = 8 

    putStrLn $ "\nTesting " ++ show numElements ++ " elements:"

    
    measureTime "Push operation" $ do
        let stack = foldl (\s x -> push x s) EmptyStack [1..numElements]
        forceEvaluation stack
        putStrLn $ "Memory usage after push: " 
                 ++ show (estimateMemorySize stack elementSize) 
                 ++ " bytes"

    
    measureTime "Pop operation" $ do
        let stack = foldl (\s x -> push x s) EmptyStack [1..numElements]
        let finalStack = foldl (\s _ -> pop s) stack [1..numElements]
        forceEvaluation finalStack


main :: IO ()
main = do
    testStackOperations 10000
    testStackOperations 20000
    testStackOperations 30000
